jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java

```diff
@@ -311,4 +356,53 @@ private void evict(boolean direct, long excess)
         }
         return oldestEntry;
     }
+
+    private static class Bucket extends Pool<RetainableByteBuffer>
+    {
+        private final int _capacity;
+
+        Bucket(int capacity, int size)
+        {
+            super(Pool.StrategyType.THREAD_ID, size, true);
+            _capacity = capacity;
+        }
+
+        @Override
+        public String toString()
+        {
+            int entries = 0;
+            int inUse = 0;
+            long used = 0;
+            long capacity = 0;
+            for (Entry entry : values())
+            {
+                entries++;
+                if (entry.isInUse())
+                {
+                    inUse++;
+
+                    // Looking at the buffer indexes is intrinsically a race. There will be some bad samples.
+                    ByteBuffer buffer = entry.getPooled().getBuffer();
+                    int pos = buffer.position();
+                    int lim = buffer.limit();
+
+                    // We can't tell the difference between empty and totally full buffers due to fill/flush mode
+                    // So we only sample usage when either pos or lim is not maximized
+                    if (pos > 0 || lim < buffer.capacity())
+                    {
+                        // If limit is not capacity, then flush mode, otherwise fill mode
+                        used += (lim < buffer.capacity()) ? lim : pos;
+                        capacity += buffer.capacity();
+                    }
+                }
+            }
+
+            return String.format("%s{capacity=%d,inuse=%d(%d%%),used=%d%%}",
```

Comment:
I'm not sure the `used` calculation will report anything meaningful, a more useful metric would be to calculate the mean 'waste' (i.e.: the difference between the requested size and the returned buffer's capacity) but that would require some extra processing.

I think we should leave this as-is and eventually revisit that later on.

Reply:
Agreed. We are kind of thwarted by the fact that we have no idea if a buffer is in fill or flush mode.
At least with this metric, it is sampling buffers that have a pointer between 0 and capacity, which is at least a hint of partial usage.